metals/src/main/scala/scala/meta/internal/metals/ReferenceProvider.scala

```diff
@@ -113,94 +113,62 @@ final class ReferenceProvider(
 
   // Returns alternatives symbols for which "goto definition" resolves to the occurrence symbol.
   private def referenceAlternatives(
-      doc: TextDocument,
-      occ: SymbolOccurrence
+      symbol: String,
+      referenceDoc: TextDocument
   ): Set[String] = {
-    val name = occ.symbol.desc.name.value
-    // Returns true if `info` is the companion object matching the occurrence class symbol.
-    def isCompanionObject(info: SymbolInformation): Boolean =
-      info.isObject &&
-        info.displayName == name &&
-        occ.symbol == Symbols.Global(
-          info.symbol.owner,
-          Descriptor.Type(info.displayName)
-        )
-    // Returns true if `info` is a named parameter of the primary constructor
-    def isContructorParam(info: SymbolInformation): Boolean = {
-      info.isParameter &&
-      info.displayName == name &&
-      occ.symbol == (Symbol(info.symbol) match {
-        case GlobalSymbol(
-              // This means it's the primary constructor
-              GlobalSymbol(owner, Descriptor.Method("<init>", "()")),
-              Descriptor.Parameter(_)
-            ) =>
-          Symbols.Global(owner.value, Descriptor.Term(name))
-        case _ =>
-          ""
-      })
+    val definitionDoc = if (referenceDoc.symbols.exists(_.symbol == symbol)) {
+      Some(referenceDoc)
+    } else {
+      for {
+        location <- definition.fromSymbol(symbol).asScala.headOption
+        source = location.getUri().toAbsolutePath
+        definitionDoc <- semanticdbs.textDocument(source).documentIncludingStale
+      } yield definitionDoc
+    }
+
+    definitionDoc match {
```

Comment:
So previously, we would check any document that the cursor currently points to. This would cause us not to find synthetic objects, since their definition only exists in the file where the case class was also defined.